### PR TITLE
[RLLib] DDPG

### PR DESCRIPTION
**Pull request by alvkao58**
Implemented DDPG Algorithm.

### CODE REVIEW

File: python/ray/rllib/ddpg/ddpg_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    def initialize(self):
+        self.sess.run(tf.global_variables_initializer())
+
+    #TODO: (not critical) Add batch normalization?
+    def sample(self, no_replay = True):
+        """Returns a batch of samples."""
+        # act in the environment, generate new samples
+        rollout = self.sampler.get_data()
+        samples = process_rollout(
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg_evaluator.py]: make sure observations + new_obs treatment is the same

File: python/ray/rllib/ddpg/ddpg_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        self.replay_buffer = ReplayBuffer(config["buffer_size"])
+        self.sampler = SyncSampler(
+                        self.env, self.model, NoFilter(),
+                        config["batch_size"], horizon=config["horizon"])
+
+    def initialize(self):
+        self.sess.run(tf.global_variables_initializer())
+
+    #TODO: (not critical) Add batch normalization?
+    def sample(self, no_replay = True):
```
Comment (ericl) [inline python/ray/rllib/ddpg/ddpg_evaluator.py]: Is it possible to remove the replay handling from DDPG, and use SyncLocalReplayOptimizer / ApexOptimizer instead? Recently we refactored DQN to have replay be handled by the optimizer classes.

File: python/ray/rllib/ddpg/replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
+from __future__ import absolute_import
```
Comment (ericl) [inline python/ray/rllib/ddpg/replay_buffer.py]: Note: a copy of these classes in in ray.rllib.optimizers already

File: python/ray/rllib/agent.py [hunk 4]
```diff
@@ -240,6 +240,9 @@ def get_agent_class(alg):
     elif alg == "PG":
         from ray.rllib import pg
         return pg.PGAgent
+    elif alg == "DDPG":
+        from ray.rllib import ddpg
+        return ddpg.DDPGAgent
```
Comment (ericl) [inline python/ray/rllib/agent.py]: Nice. We should also make an issue to update the docs; there are a couple new algs not documented.

File: python/ray/rllib/ddpg/ddpg_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        self.actor_grads = tf.gradients(self.model.actor_loss, self.model.actor_var_list)
+        a_grads_and_vars = list(zip(self.actor_grads, self.model.actor_var_list))
+        a_opt = tf.train.AdamOptimizer(self.config["actor_lr"])
+        self._apply_a_gradients = a_opt.apply_gradients(a_grads_and_vars)
+
+    def compute_gradients(self, samples):
+        """ Returns gradient w.r.t. samples."""
+        # actor gradients
+        actor_actions = self.sess.run(self.model.output_action,
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg_evaluator.py]: don't you track the actor samples in `samples["actions"]` or something?

Comment (alvkao58) [reply python/ray/rllib/ddpg/ddpg_evaluator.py]: yeah, but looking at https://arxiv.org/pdf/1509.02971.pdf, aren't the actions according to what the actor currently outputs?

Comment (richardliaw) [reply python/ray/rllib/ddpg/ddpg_evaluator.py]: ok yeah good point

**Review by richardliaw — CHANGES_REQUESTED** [review]
some comments

**Comment by alvkao58** [comment]
Plot of average reward over last 10 episodes (episodes are 200 steps) for Pendulum-v0. 
learninggraph

Plot shows 80 episodes in total, average reward starts noticeably increasing at around 60 episodes.

**Review by richardliaw — CHANGES_REQUESTED** [review]

File: python/ray/rllib/utils/sampler.py [hunk 6]
```diff
@@ -20,7 +20,7 @@ class PartialRollout(object):
         last_r (float): Value of next state. Used for bootstrapping.
     """
 
-    fields = ["observations", "actions", "rewards", "terminal", "features"]
+    fields = ["obs", "actions", "rewards", "new_obs", "dones", "features"]
```
[the remark this answers is not in the document]
Comment (alvkao58) [reply python/ray/rllib/utils/sampler.py]: I did it for convenience because LocalSyncReplayOptimizer uses "obs", "new_obs", and "dones". I can change it back if necessary.

File: python/ray/rllib/ddpg/ddpg.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            stats = [stats]
+
+        mean_10ep_reward = 0.0
+        mean_10ep_length = 0.0
+        num_episodes = 0
+
+        for s in stats:
+            mean_10ep_reward += s["mean_10ep_reward"] / len(stats)
+            mean_10ep_length += s["mean_10ep_length"] / len(stats)
+            num_episodes += s["num_episodes"]
```
[the remark this answers is not in the document]
Comment (alvkao58) [reply python/ray/rllib/ddpg/ddpg.py]: Is episode_reward_mean in TrainingResult meant to give an average from the start of time?

File: python/ray/rllib/ddpg/ddpg_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        self.registry = registry
+        self.env = ModelCatalog.get_preprocessor_as_wrapper(
+            registry, env_creator(config["env_config"]))
+        self.config = config
+
+        self.sess = tf.Session()
+
+        with tf.variable_scope("model"):
+            self.model = DDPGModel(self.registry,
+                                   self.env,
```
[the remark this answers is not in the document]
Comment (alvkao58) [reply python/ray/rllib/ddpg/ddpg_evaluator.py]: I think that if I don't do this, the model and target_model end up sharing weights?

File: python/ray/rllib/ddpg/ddpg.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/ddpg/ddpg.py]: It will give an average over the number of episodes finished since the last TrainingResult.

I think here we should just do a lot of optimizer steps and return a TrainingResult only after a while.

**Comment by alvkao58** [comment]
retest please

**Comment by richardliaw** [comment]
retest this please

File: python/ray/rllib/utils/sampler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alvkao58) [reply python/ray/rllib/utils/sampler.py]: I made this consistent across all algorithms that use this sampler.

File: python/ray/rllib/ddpg/random_process.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+
+import numpy as np
```
[the remark this answers is not in the document]
Comment (alvkao58) [reply python/ray/rllib/ddpg/random_process.py]: OK, moved reference to top of file.

On Pendulum, it's hard to tell if the random process makes a difference or not.

**Comment by alvkao58** [comment]
Sample reward graph after changes.

reward

**Review by richardliaw — CHANGES_REQUESTED** [review]

**Comment by richardliaw** [comment]
The graph is without batchnorm?

**Comment by alvkao58** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by alvkao58** [comment]
retest this please

**Comment by alvkao58** [comment]
retest this please

File: python/ray/rllib/ddpg/ddpg_evaluator.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    def sample(self):
+        """Returns a batch of samples."""
+
+        rollout = self.sampler.get_data()
+        rollout.data["weights"] = np.ones_like(rollout.data["rewards"])
+
+        samples = process_rollout(
+                    rollout, NoFilter(),
+                    gamma=1.0, use_gae=False)
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg_evaluator.py]: nit: we should make a comment as to why this gamma (or should not be the `gamma` from the config)

File: python/ray/rllib/ddpg/ddpg.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        episode_rewards = []
+        episode_lengths = []
+        metric_lists = [a.get_completed_rollout_metrics.remote()
+                        for a in self.remote_evaluators]
+        for metrics in metric_lists:
+            for episode in ray.get(metrics):
+                episode_lengths.append(episode.episode_length)
+                episode_rewards.append(episode.episode_reward)
+        avg_reward = (
+            np.mean(episode_rewards) if episode_rewards else float('nan'))
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg.py]: nit: I believe you can just do `np.mean(episode_rewards)` and `np.sum` (below)to the same effect without the if-else cases

File: python/ray/rllib/ddpg/ddpg.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    "noise_parameters": {
+        "mu": 0,
+        "sigma": 0.2,
+        "theta": 0.15,
+    },
+
+    # Number of local steps taken for each call to sample
+    "num_local_steps": 1,
+    # Number of workers (excluding master)
+    "num_workers": 1,
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg.py]: what happens when num_workers is 0? does the algorith still output anything?

File: python/ray/rllib/ddpg/models.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        """Sets up actor network."""
+        with tf.variable_scope("actor", reuse=tf.AUTO_REUSE):
+            self.actor_network = DDPGActor(
+                                self.obs, self.ac_size,
+                                options={"action_bound": self.action_bound})
+            self.output_action = self.actor_network.outputs
+
+    def get_weights(self):
+        """Returns critic weights, actor weights."""
+        return self.critic_vars.get_weights(), self.actor_vars.get_weights()
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/models.py]: nit: it would be great if we could just have this as one return dict

File: python/ray/rllib/agent.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/rllib/agent.py]: @alvkao58 can you take care of these?

**Review by richardliaw — COMMENTED** [review]
OK this is looking really good! We have one last thing, which is to add this to our test suite in `multi_node_tests.sh` and add an example to the regression tests folder. 

Otherwise, this is ready to merge (and will probably be done after #1868)

File: python/ray/rllib/ddpg/ddpg.py (hunk 12 above)
Comment (alvkao58) [reply python/ray/rllib/ddpg/ddpg.py]: As of now, it doesn't output anything when num_workers is 0, but I could update the stats so it takes metrics from the local evaluator when there's no remote evaluators.

File: python/ray/rllib/ddpg/models.py (hunk 13 above)
Comment (alvkao58) [reply python/ray/rllib/ddpg/models.py]: Correct me if I'm wrong, but I think that would make setting weights more annoying?

File: python/ray/rllib/ddpg/ddpg.py (hunk 12 above)
Comment (richardliaw) [reply python/ray/rllib/ddpg/ddpg.py]: OK let's do that

File: python/ray/rllib/agent.py (hunk 4 above)
Comment (alvkao58) [reply python/ray/rllib/agent.py]: any particular sanity checks you want to see added to multi_node_tests.sh?

File: python/ray/rllib/ddpg/models.py (hunk 13 above)
Comment (richardliaw) [reply python/ray/rllib/ddpg/models.py]: That's true but only slightly; the benefit of doing this is to that different optimizers can use the DDPG evaluator (which is one of the main points of RLlib)

File: python/ray/rllib/agent.py (hunk 4 above)
Comment (alvkao58) [reply python/ray/rllib/agent.py]: any particular sanity checks you want to see added to multi_node_tests.sh?

File: python/ray/rllib/ddpg/models.py (hunk 13 above)
Comment (richardliaw) [reply python/ray/rllib/ddpg/models.py]: OK let's leave this as is and fix it in a later PR if necessary.

File: python/ray/rllib/__init__.py [hunk 14]
```diff
@@ -8,8 +8,8 @@
 
 
 def _register_all():
-    for key in ["PPO", "ES", "DQN", "APEX", "A3C", "BC", "PG", "__fake",
-                "__sigmoid_fake_data", "__parameter_tuning"]:
+    for key in ["PPO", "ES", "DQN", "APEX", "A3C", "BC", "PG", "DDPG",
+                "__fake", "__sigmoid_fake_data", "__parameter_tuning"]:
```
Comment (ericl) [inline python/ray/rllib/__init__.py]: Let's rename this to DDPG3

**Review by richardliaw — APPROVED** [review]
